dev/ProgressBar/ProgressBar.cpp

```diff
@@ -0,0 +1,149 @@
+
+    UpdateStates();
+}
+
+void ProgressBar::OnShowPausedPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
+}
+
+void ProgressBar::OnShowErrorPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
+}
+
+void ProgressBar::UpdateStates()
+{
+    m_sizeChangedRevoker.revoke();
+
+    if (ShowError())
+    {
+        winrt::VisualStateManager::GoToState(*this, s_ErrorStateName, true);
+    }
+    else if (ShowPaused() && IsIndeterminate())
+    {
+        winrt::VisualStateManager::GoToState(*this, s_ErrorStateName, true); // Paused-Indeterminate state same visual treatment as Error state
+    }
+    else if (ShowPaused())
+    {
+        winrt::VisualStateManager::GoToState(*this, s_PausedStateName, true);
+    }
+    else if (IsIndeterminate())
+    {
+        UpdateWidthBasedTemplateSettings(nullptr, nullptr);
+        winrt::VisualStateManager::GoToState(*this, s_IndeterminateStateName, true);
+        m_sizeChangedRevoker = this->SizeChanged(winrt::auto_revoke, { this, &ProgressBar::UpdateWidthBasedTemplateSettings });
+    }
+    else if (!IsIndeterminate())
+    {
+        winrt::VisualStateManager::GoToState(*this, s_DeterminateStateName, true);
+    }
+}
+
+void ProgressBar::SetProgressBarIndicatorWidth()
+{
+    if (auto&& progressBar = m_layoutRoot.get())
+    {
+        if (auto&& progressBarIndicator = m_progressBarIndicator.get())
+        {
+            const double progressBarWidth = progressBar.ActualWidth();
+            const double maximum = Maximum();
+            const double minimum = Minimum();
+
+            if (std::abs(maximum - minimum) > DBL_EPSILON)
+            {
+                const double increment = progressBarWidth / (maximum - minimum);
+                progressBarIndicator.Width(increment * (Value() - minimum));
+            }
+            else
+            {
+                progressBarIndicator.Width(0); // Error
+            }
+        }
+    }
+}
+
+void ProgressBar::UpdateWidthBasedTemplateSettings(const winrt::IInspectable&, const winrt::IInspectable&)
+{
+    const auto templateSettings = winrt::get_self<::ProgressBarTemplateSettings>(TemplateSettings());
+
+    if (auto&& progressBarIndicator = m_progressBarIndicator.get())
+    {
+        auto const [width, height] = [progressBar = m_layoutRoot.get()]()
+        {
+            if (progressBar)
+            {
+                const float width = static_cast<float>(progressBar.ActualWidth());
+                const float height = static_cast<float>(progressBar.ActualHeight());
+                return std::make_tuple(width, height);
+            }
+            return std::make_tuple(0.0f, 0.0f);
+        }();
+
+        progressBarIndicator.Width(width / 3);
+
+        templateSettings->ContainerAnimationEndPosition(width);
+
+        const winrt::Windows::UI::Xaml::Media::RectangleGeometry rectangle = [width, height, padding = Padding()]()
```

Comment:
>winrt::Windows::UI::Xaml::Media::RectangleGeometry [](start = 14, length = 50)

I think you should make this auto instead of a specified type.

Reply:
This class could not be initialized with `auto`

Reply:
if nothing else, use the shorter type names like "winrt::RectangleGeometry", as you've done in the body of the lambda.

Reply:
And did you get a compile error about the use of auto? If so could you provide it here?

Reply:
![image](https://user-images.githubusercontent.com/7976322/67587420-b3367380-f708-11e9-8dc8-082bbeb53659.png)